src/onesided.jl

```diff
@@ -143,42 +143,38 @@ function Win_unlock(rank::Integer, win::Win)
     @mpichk ccall((:MPI_Win_unlock, libmpi), Cint, (Cint, MPI_Win), rank, win)
 end
 
-function Get(origin_buffer, count::Integer, target_rank::Integer, target_disp::Integer, win::Win)
-    T = eltype(origin_buffer)
+
+# TODO: add some sort of "remote buffer": a way to specify different datatypes/counts
+
+function Get(origin_buf::Buffer, target_rank::Integer, target_disp::Integer, win::Win)
     # int MPI_Get(void *origin_addr, int origin_count,
     #             MPI_Datatype origin_datatype, int target_rank,
     #             MPI_Aint target_disp, int target_count,
     #             MPI_Datatype target_datatype, MPI_Win win)
     @mpichk ccall((:MPI_Get, libmpi), Cint,
                   (MPIPtr, Cint, MPI_Datatype, Cint, Cptrdiff_t, Cint, MPI_Datatype, MPI_Win),
-                  origin_buffer, count, Datatype(T), target_rank, Cptrdiff_t(target_disp), count, Datatype(T), win)
-end
-function Get(origin_buffer::AbstractArray{T}, target_rank::Integer, win::Win) where T
-    count = length(origin_buffer)
-    Get(origin_buffer, count, target_rank, 0, win)
+                  origin_buf.data, origin_buf.count, origin_buf.datatype,
+                  target_rank, Cptrdiff_t(target_disp), origin_buf.count, origin_buf.datatype, win)
 end
-function Get(origin_value::Ref{T}, target_rank::Integer, win::Win) where T
-    Get(origin_value, 1, target_rank, 0, win)
+function Get(origin::Union{AbstractArray,Ref}, target_rank::Integer, win::Win)
+    Get(Buffer(origin), target_rank, 0, win)
 end
 
-function Put(origin_buffer, count::Integer, target_rank::Integer, target_disp::Integer, win::Win)
+function Put(origin_buf::Buffer, count::Integer, target_rank::Integer, target_disp::Integer, win::Win)
     # int MPI_Put(const void *origin_addr, int origin_count,
     #             MPI_Datatype origin_datatype, int target_rank,
     #             MPI_Aint target_disp, int target_count,
     #             MPI_Datatype target_datatype, MPI_Win win)
-    T = eltype(origin_buffer)
     @mpichk ccall((:MPI_Put, libmpi), Cint,
                   (MPIPtr, Cint, MPI_Datatype, Cint, Cptrdiff_t, Cint, MPI_Datatype, MPI_Win),
-                  origin_buffer, count, Datatype(T), target_rank, Cptrdiff_t(target_disp), count, Datatype(T), win)
+                  origin_buf.data, origin_buf.count, origin_buf.datatype,
+                  target_rank, Cptrdiff_t(target_disp), origin_buf.count, origin_buf.datatype, win)
```

Comment:
I notice that `count::Integer` is passed as an argument but not used in the new method definition. It should probably be removed.

Reply:
Thanks!